vm/devices/serial/serial_debugcon/src/lib.rs

```diff
@@ -0,0 +1,193 @@
+use std::task::ready;
+use std::task::Context;
+use std::task::Poll;
+use std::task::Waker;
+use vmcore::device_state::ChangeDeviceState;
+
+/// A Debugcon serial port emulator.
+#[derive(InspectMut)]
+pub struct SerialDebugcon {
+    // Fixed configuration
+    io_port: u16,
+    #[inspect(skip)]
+    io_region: (&'static str, RangeInclusive<u16>),
+
+    // Runtime glue
+    #[inspect(mut)]
+    io: Box<dyn SerialIo>,
+
+    // Volatile state
+    #[inspect(with = "VecDeque::len")]
+    tx_buffer: VecDeque<u8>,
+    #[inspect(skip)]
+    tx_waker: Option<Waker>,
+}
+
+impl SerialDebugcon {
+    /// Returns a new emulator instance.
+    pub fn new(port: u16, io: Box<dyn SerialIo>) -> Self {
+        Self {
+            io_port: port,
+            io_region: ("debugcon", port..=port),
+            io,
+            tx_buffer: VecDeque::new(),
+            tx_waker: None,
+        }
+    }
+
+    /// Synchronize interrupt and waker state with device state.
+    fn sync(&mut self) {
+        // Wake to poll if there are any bytes to write.
+        if !self.tx_buffer.is_empty() {
+            if let Some(waker) = self.tx_waker.take() {
+                waker.wake();
+            }
+        }
+    }
+
+    fn poll_tx(&mut self, cx: &mut Context<'_>) -> Poll<()> {
+        while !self.tx_buffer.is_empty() {
+            let (buf, _) = self.tx_buffer.as_slices();
+            match ready!(Pin::new(&mut self.io).poll_write(cx, buf)) {
+                Ok(n) => {
+                    assert_ne!(n, 0);
+                    self.tx_buffer.drain(..n);
+                }
+                Err(err) if err.kind() == ErrorKind::BrokenPipe => {
+                    tracing::info!("debugcon output broken pipe");
+                }
+                Err(err) => {
+                    tracelimit::error_ratelimited!(
+                        len = buf.len(),
+                        error = &err as &dyn std::error::Error,
+                        "debugcon write failed, dropping data"
+                    );
+                    self.tx_buffer.drain(..buf.len());
+                }
+            }
+        }
+        // Wait for more bytes to write.
+        self.tx_waker = Some(cx.waker().clone());
+        Poll::Pending
+    }
+}
+
+impl ChangeDeviceState for SerialDebugcon {
+    fn start(&mut self) {}
+
+    async fn stop(&mut self) {}
+
+    async fn reset(&mut self) {
+        self.tx_buffer.clear();
+    }
+}
+
+impl ChipsetDevice for SerialDebugcon {
+    fn supports_pio(&mut self) -> Option<&mut dyn PortIoIntercept> {
+        Some(self)
+    }
+
+    fn supports_poll_device(&mut self) -> Option<&mut dyn PollDevice> {
+        Some(self)
+    }
+}
+
+impl PollDevice for SerialDebugcon {
+    fn poll_device(&mut self, cx: &mut Context<'_>) {
+        let _ = self.poll_tx(cx);
+        self.sync();
+    }
+}
+
+impl PortIoIntercept for SerialDebugcon {
+    fn io_read(&mut self, io_port: u16, data: &mut [u8]) -> IoResult {
+        if io_port != self.io_port {
```

Comment:
How could this happen?

Reply:
Just an eager sanity check, akin to a `_ => return Err` block in more complex devices. We can nix it.

Reply:
Either way.